### PR TITLE
Frogger-game

### PR DESCRIPTION
**Pull request by zhenyakornilov**
### Object-Oriented JS - Frogger-game

[Demo](https://zhenyakornilov.github.io/kottans-frogger-game-oop/) |
[Code base](https://github.com/zhenyakornilov/kottans-frogger-game-oop)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

**Comment by zhenyakornilov** [comment]
### Covered some common mistakes:
1. Class **Enemy** do not refer to global **player** variable, **player** passed to **Enemy** constructor instead.
2. **Enemy** **update()** method do not call **checkOverlap()** method inside. Moved **checkOverlap()** method to **updateEntities** function in **engine.js** instead.

**Comment by OleksiyRudenko** [comment]
> 2\. **Enemy** **update()** method do not call **checkOverlap()** method inside. Moved **checkOverlap()** method to **updateEntities** function in **engine.js** instead.

Please bring it back to the `app.js` for the sake of code review scope. It can perfectly be a part of Enemy class (as it is specific to Enemy behaviour), just needs to be a method on its own.

File: submissions/zhenyakornilov/frogger-game/app.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    this.speed = speed;
+    this.direction = direction;
+    this.player = player;
+  }
+
+  update(dt) {
+    if (this.direction === "left") {
+      this.x -= dt * this.speed;
+      if (this.x < 0) {
+        this.x = GAME_BOARD_WIDTH + 10;
```
Comment (OleksiyRudenko) [inline submissions/zhenyakornilov/frogger-game/app.js:44]: WHat is `10` here and across the codebase? I guess `-10` is the same but just negative. If true then having `OFFSET = 10` and referring to it as `OFFSET` and `-OFFSET` would be fine.
Of course `OFFSET` word doesn't explain it application, so invent a better name.

File: submissions/zhenyakornilov/frogger-game/app.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      this.x += dt * this.speed;
+      if (this.x > GAME_BOARD_WIDTH) {
+        this.x = -10;
+      }
+    }
+  }
+
+  checkOverlap() {
+    if (
+      this.player.x < this.x + 60 &&
```
Comment (OleksiyRudenko) [inline submissions/zhenyakornilov/frogger-game/app.js:56]: What is `60`?

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@zhenyakornilov very well coded!

Apart of bringing collision checking function back as mentioned above, let's add more semantics to the constants so that peer developers were able to **read** what every number means and implement changes quickly when needed.

**Comment by zhenyakornilov** [comment]
Made some changes

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@zhenyakornilov please go through your code and extrapolate previous comments re magic numbers to all other similar situations in your code.
Do not expect a reviewer to "red pen" every single tiny bug in the code. Own it truly.

**Comment by zhenyakornilov** [comment]
It seems to be a bit better now.

**Review by OleksiyRudenko — APPROVED** [review]
@zhenyakornilov a really good job!

Hint on naming in graphic environments:
In regular structures like matrices, game fields (which are also matrices) elements' coordinates are defined by row (or column), matrix row height (or matrix column width), and offset.
So, for e.g. `y = verticalOffset + row * rowHeight`.  Items to place into the matrix may have their own individual offsets (like in this game where visual part has offsets against its pixel representation).
So having defined offsets, widths and heights it is possible to calculate any coordinates based on row and column index.

**Pull request merged**